File: app/src/test/java/org/mifos/mobilebanking/BeneficiaryListPresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.beneficiary.Beneficiary;
import org.mifos.mobilebanking.presenters.BeneficiaryListPresenter;
import org.mifos.mobilebanking.ui.views.BeneficiariesView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.List;


import io.reactivex.Observable;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class BeneficiaryListPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    BeneficiariesView view;

    private BeneficiaryListPresenter presenter;
    private List<Beneficiary> beneficiaryList;

    @Before
    public void setUp() throws Exception {
        presenter = new BeneficiaryListPresenter(dataManager, context);
        presenter.attachView(view);

        beneficiaryList = FakeRemoteDataSource.getBeneficiaries();
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testLoadBeneficiaries() throws Exception {
        when(dataManager.getBeneficiaryList()).thenReturn(Observable.just(beneficiaryList));

        presenter.loadBeneficiaries();

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showBeneficiaryList(beneficiaryList);
        verify(view, never()).showError(context.getString(R.string.error_fetching_beneficiaries));
    }

    @Test
    public void testLoadBeneficiariesFails() throws Exception {
        when(dataManager.getBeneficiaryList()).thenReturn(Observable.<List<Beneficiary>>error(new
                RuntimeException()));

        presenter.loadBeneficiaries();

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showError(context.getString(R.string.error_fetching_beneficiaries));
        verify(view, never()).showBeneficiaryList(beneficiaryList);
    }

}

File: app/src/test/java/org/mifos/mobile/BeneficiaryListPresenterTest.kt
package org.mifos.mobile

import android.content.Context
import io.reactivex.Observable

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.beneficiary.Beneficiary
import org.mifos.mobile.presenters.BeneficiaryListPresenter
import org.mifos.mobile.ui.views.BeneficiariesView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner::class)
class BeneficiaryListPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: BeneficiariesView? = null
    private var presenter: BeneficiaryListPresenter? = null
    private var beneficiaryList: List<Beneficiary?>? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = BeneficiaryListPresenter(dataManager!!, context!!)
        presenter?.attachView(view)
        beneficiaryList = FakeRemoteDataSource.beneficiaries
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(Exception::class)
    fun testLoadBeneficiaries() {
        Mockito.`when`(dataManager?.beneficiaryList).thenReturn(Observable.just(beneficiaryList))
        presenter?.loadBeneficiaries()
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showBeneficiaryList(beneficiaryList)
        Mockito.verify(view, Mockito.never())?.showError(context?.getString(R.string.error_fetching_beneficiaries))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadBeneficiariesFails() {
        Mockito.`when`(dataManager?.beneficiaryList).thenReturn(Observable.error(RuntimeException()))
        presenter?.loadBeneficiaries()
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showError(context?.getString(R.string.error_fetching_beneficiaries))
        Mockito.verify(view, Mockito.never())?.showBeneficiaryList(beneficiaryList)
    }
}